pkg/flatten/flatten.go

```diff
@@ -0,0 +1,268 @@
+//
+// Copyright (c) 2019-2021 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package flatten
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+
+	devfile "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/api/v2/pkg/utils/overriding"
+	"github.com/devfile/library/pkg/flatten/network"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ResolverTools contains required structs and data for resolving remote components of a devfile (plugins and parents)
+type ResolverTools struct {
+	// DefaultNamespace is the default namespace to use for resolving Kubernetes ImportReferences that do not include one
+	DefaultNamespace string
+	// DefaultRegistryURL is the default registry URL to use when a component specifies an id but not registryURL
+	DefaultRegistryURL string
+	// Context is the context used for making Kubernetes or HTTP requests
+	Context context.Context
+	// K8sClient is the Kubernetes client instance used for interacting with a cluster
+	K8sClient client.Client
+	// HttpClient is the HTTP client used for making network requests when resolving plugins or parents.
+	HttpClient network.HTTPGetter
+}
+
+// ResolveDevWorkspace takes a DevWorkspaceTemplateSpec and returns a "resolved" version of it -- i.e. one where all plugins and parents
+// are inlined as components.
+// TODO:
+// - Implement flattening for DevWorkspace parents
```

Comment:
I see `resolveParentComponent` func, isn't it already done?

Reply:
Forgot to remove the TODO after implementing :)